wip: add Ruby project management lesson (#27128)
TheOdinProject/curriculum: ruby/object_oriented_programming_basics/managing_ruby_projects.md

### Introduction

Perhaps going through the curriculum you were wondering about how programmers structure their projects. Do they really just put their entire code in a single file? If they don't, how are they able to use code across multiple files? Are there agreed-upon rules? Common patterns? What are the Ruby ways of doing things? Maybe you haven't yet thought about that but very shortly knowing how to structure your project is going to help you and others navigate the code you've written.

### Lesson overview

This section contains a general overview of topics that you will learn in this lesson.
- The benefits of splitting your code into separate files.
- How to make use of code from a separate file.
- Namespace and scope of `require`d files.
- What development dependencies are.
- What gems are.
- What Bundler is and why it is useful.
- What `bundle init` does and what a `Gemfile` is.
- How to get the most out of `Ruby LSP` VSCode extension.

### Confusion, convention, convenience

> Recall how you've worked with the projects in Foundations: the HTML, CSS and JS all lived in separate files, becoming "one" thanks to linking CSS and JS in the HTML file(s). Being able to divide the type of work across multiple files hopefully organized your work better and prevented confusion. Moving forward you can imagine that your files are only going to get longer and more plentiful in number, thus making having some kind of structure necessary. Additionally, you might've already realized how useful git tracking changes per-file together with GitHub is by making it a lot easier to track down a bug if you know that your last commit only affected your JS file.

Reviewer: Love how you connected this to foundations stuff!

Author: Good shows have callbacks for folks that watched the previous seasons. Maybe that works for curriculums too? 😆

Both of those approaches are going to execute the file, allowing you to use their contents. If you try to require something for the second time, nothing will happen and the requires will return `false`.

Convention is that `require_relative` is used for your own code, while `require` is used for things outside of it, like gems that your app depend on. Now, there's couple of things that are important to know about requiring files in Ruby. First is that all required code is put into the same namespace. This means that if you have the same names for methods, modules, classes and so on they will be added together in order they were required. For example:
```rb
# all files are in the same directory for simplicity's sake

# foo.rb
def foo(bar)

>   `{#} is awesome!`

Author: Whoops. I'll fix that, apologies!

RubyGems has been part of Ruby standard library since version 1.9 and is used to get those amazing gems onto your computer. Remember that bit about `require` going through installed gems to potentially find a file you're looking for? That's the work of RubyGems. Another cool part about it? It is a gem itself! Let's give it a try.

Create a new Ruby file `main.rb` in a directory called `colorful`:
```rb
require 'colorize'

puts 'Red goes faster!'.colorize(:red)

puts "I'm blue da ba dee da ba di!".colorize(:blue)

puts "It ain't easy bein' green...".colorize(:green)
```
You're probably itching to see all those colours, so run your file with `ruby foo.rb` to see them... or rather, a LoadError. Right - you need to install that gem first! Do that with `gem install colorize` and you'll see RubyGems in action. Your system now has access to the `Colorize` gem!

> Your system now has access to the `Colorize` gem - what about others that would like to use your code? Yeah, they would need to `gem install` it as well, no big deal. But what if you have dozens of gems? How do you make sure that the versions you use are the versions others download? This sounds rather tedious. Enter: Bundler. Another gem, part of RubyGems but released independently. Bundler allows you to declare what gems your project needs - down to their version. As for others, Bundler allows them to take that declaration, a simple file called `Gemfile` and use it to install those gems in a quick `bundle install`.  Since gem installs are global, you need a way to run only those particular gem versions that are declared in the `Gemfile`. You can do that by using `bundle exec` followed by a command that you want to execute - most likely `bundle exec ruby foo.rb`.

Author: It sounded more poetic in my head, that repetition. I'll do something with it, thanks!

Let's make sure whoever wants to use our script can do so:
```bash
bundle init
bundle add colorize
```
This has created two files: `Gemfile` and `Gemfile.lock`. Let's take a look into both:
```rb
# Gemfile

# frozen_string_literal: true

source "https://rubygems.org"

# gem "rails"

gem "colorize", "~> 1.1"

# Gemfile.lock
GEM
  remote: https://rubygems.org/
  specs:
    colorize (1.1.0)

PLATFORMS
  ruby

>   x86_64-linux

Author: I guess to be on the safe side I can add a small note. In my mind this is self-explanatory due to being under "PLATFORMS" but that term might not bring up the same associations that I get.

DEPENDENCIES
  colorize (~> 1.1)

BUNDLED WITH
   2.5.4
```

> There's not much in those but as you can see, the `Gemfile` has information on where to get the gems from and what gems are required. The `"~> 1.1"` is something called a version constraint, particularly one called pessimistic constraint. It relies on semantic versioning - basically, the first number is the major version, second is the minor version and a third, if it exists is the patch number. Major versions can break things from previous versions - think changing method names for example. Minor versions can add and change things but they can't break anything. Patches happen when you introduce bug fixes that don't break anything. So, if people behind a gem maintain it in line with semantic versioning, you can rely on this pessimistic constraint never letting your project have a gem version that could potentially break your app - it is equivalent to `gem "colorize", ">= 1.1", "<2.0"`.

Author: The problem I found is that if I want to explain the syntax, there's no way of avoiding the semver, because it depends on that. I guess that I could say something like: 
```suggestion
There's not much in those but as you can see, the `Gemfile` has information on where to get the gems from and what gems are required. The `"~> 1.1"` is something called a version constraint, particularly one called pessimistic constraint. It relies on semantic versioning - in short, it is supposed to let your gem be updated but only until breaking changes have been introduced. It is equivalent to `gem "colorize", ">= 1.1", "<2.0"`.
```
and throw the rubygem's explanation in Additional Resources.

`Gemfile.lock` has information on what was the last environment that should be able to run your app. Bundler will use it to install the same versions of gems even if `Gemfile` could potentially allow for newer versions to be installed.

### .ruby-version

There's another one small bit that is important to give to folks that will run your code: the target Ruby version of your project. You can do it easily by running `rbenv local 3.2.2` as it creates a `.ruby-version` file with the version declared. This is recognized by many other tools to figure out what Ruby version your project is running - for example, `rbenv` will no longer use the `global` Ruby version and the `Ruby LSP` VSCode extension will also change its behavior.

### Ruby LSP in VSCode

> Earlier in the curriculum you've been instructed to choose the `No - launch without bundle` option when `Ruby LSP` told you about not finding a `Gemfile` - that also resulted with an error concerning `RuboCop`. In the next lesson we are going to go over `RuboCop` and how it should land in your project's `Gemfile`. After that, you are going to enjoy all the benefits of using `Ruby LSP` and its `RuboCop` integration and your projects are going to be set up like a real pro.

Author: Apologies, no, it doesn't but I want it to be part of this PR. I decided to ask for opinion already because I wasn't sure about the quality of the lesson. I'll add the modified bits of previous lessons soon TM.

> Earlier in the curriculum you've been instructed to choose the `No - launch without bundle` option when `Ruby LSP` told you about not finding a `Gemfile` - that also resulted with an error concerning `RuboCop`. In the next lesson we are going to go over `RuboCop` and how it should land in your project's `Gemfile`. After that, you are going to enjoy all the benefits of using `Ruby LSP` and its `RuboCop` integration and your projects are going to be set up like a real pro.

Author: Added with [1dbff49](https://github.com/TheOdinProject/curriculum/pull/27128/commits/1dbff4924d758acbab2e6c08914ee75afe4cb5fb). I also took the liberty to remove the fluff about finding a different extension for different IDEs as the curriculum doesn't support anything but VSCode + it's rather self-explanatory that one should look for extensions for their IDE [bfbf12e](https://github.com/TheOdinProject/curriculum/pull/27128/commits/bfbf12ebff34774947ad597869434d2131d365bf).

> Earlier in the curriculum you've been instructed to choose the `No - launch without bundle` option when `Ruby LSP` told you about not finding a `Gemfile` - that also resulted with an error concerning `RuboCop`. In the next lesson we are going to go over `RuboCop` and how it should land in your project's `Gemfile`. After that, you are going to enjoy all the benefits of using `Ruby LSP` and its `RuboCop` integration and your projects are going to be set up like a real pro.

Author: Had to update the wording here since in the meantime the message has changed which was what I used for `Installing Ruby` lesson but forgot to change it here.

### Assignment

1. Read [Eric Mathison's Understanding require and Friends in Ruby](https://github.com/ericmathison/articles/blob/master/understaning-require-in-ruby.md).
2. Check out the [RubyGems Basics](https://guides.rubygems.org/rubygems-basics/).

> 3. Skim through [Patterns guide from RubyGems](https://guides.rubygems.org/patterns/), paying closer attention to `Semantic versioning`, `Declaring dependencies` and `Loading code`.

Author: Additional Resources. Please let it live there 😆

### Knowledge check

This section contains questions for you to check your understanding of this lesson on your own. If you’re having trouble answering a question, click it and review the material it links to.
- [A KNOWLEDGE CHECK QUESTION](A-KNOWLEDGE-CHECK-URL)

### Additional resources

This section contains helpful links to related content. It isn’t required, so consider it supplemental.
- https://www.youtube.com/watch?v=j2V-A8vvLP0 Andre Arko talk
- https://www.youtube.com/watch?v=I0a5zv7uBHw Adam McCrea talk
- https://www.youtube.com/watch?v=fkfCXFxJW08 Jenny Shen talk

https://medium.com/@connorstack/understanding-ruby-load-require-gems-bundler-and-rails-autoloading-from-the-bottom-up-3b422902ca0
https://thoughtbot.com/blog/following-the-path
https://ryanbigg.com/2017/11/how-require-loads-a-gem
https://github.com/ericmathison/articles/blob/master/understaning-require-in-ruby.md
https://andre.arko.net/2015/04/28/how-does-bundler-work-anyway/
https://www.justinweiss.com/articles/how-do-gems-work/
https://bundler.io/v2.5/man/gemfile.5.html
https://guides.rubygems.org/patterns/

> https://semver.org/

Author: This is very WIP 😆 I more-or-less dropped resources that I've used to create the lesson and figured I'll think about what should be put there after we wrap most of the work up. I *think* I want to choose one of these three talks, I watched the Andre Arko one in its entirety and watched some of Jenny Shen's and while Jenny's is more fresh, I think it takes a little too long to get the ball rolling.

Now, it'd be nice to be able to also divide your Ruby work across multiple files since you're about to set out onto the more complicated OOP projects, where many distinct parts are going to work in unison. Not only is that possible, there are also some conventions around Ruby project structure that aim at making it easier for you and others to work with and navigate your project.

### Making use of multiple files

If you are to split your code across multiple files, you first will need to know how to make sure code from one file and be used in another file. There are two main ways to do that:

- `require_relative` which takes in a string of a relative path, like `require_relative 'lib/bogo_sort'` is going to look for `lib/bogo_sort.rb` relative to the location of the current file,

> - `require` is trickier - there are two ways of using it: one is providing a relative path which is recognized by using `../` or `./` at the start or an absolute path. This works like `require_relative`. Another way is by providing a file name without the extension, like `require 'csv'` is going to look for a `csv.rb` in the Ruby's `$LOAD_PATH` global variable which by default contains the Ruby standard library. There are other file extensions it might look for but this is not important at this point. In addition to that, if it doesn't find that file in `$LOAD_PATH` it is going to look through installed gems (more on those later) to see if the file's there.

Author: Need to get acquainted with the rules governing commas, I only know of Oxford comma really 😆 In Polish we use *a lot* of punctuation, I've struggled with using too many commas for very long so nowadays I just don't do them at all. Besides this comment. Right.

> - `require` is trickier - there are two ways of using it: one is providing a relative path which is recognized by using `../` or `./` at the start or an absolute path. This works like `require_relative`. Another way is by providing a file name without the extension, like `require 'csv'` is going to look for a `csv.rb` in the Ruby's `$LOAD_PATH` global variable which by default contains the Ruby standard library. There are other file extensions it might look for, but this is not important at this point. In addition to that, if it doesn't find that file in `$LOAD_PATH` it is going to look through installed gems (more on those later) to see if the file is there.

Author: Hmm... yeah. I'll need to come back here too to clarify this. It looks like I made an error defining the two `require`s which is just... sad.

Both of those approaches are going to execute the file, allowing you to use their contents. If you try to require something for the second time, nothing will happen, and the requires will return `false`.

> Convention is that `require_relative` is used for your own code, while `require` is used for things outside of it, like gems that your app depend on. Now, there's couple of things that are important to know about requiring files in Ruby. <span id="namespace">First is that all required code is put into the same namespace. This means that if you have the same names for methods, modules, classes and so on they will be added together in order they were required.</span> For example:

Author: I looked into how Bundler gem itself is set up:
https://github.com/rubygems/rubygems/blob/5cedfc8ad923cbc679049ed69fbb1baf1753617f/bundler/lib/bundler.rb#L3-L15

Here are the tests:
https://github.com/rubygems/rubygems/blob/5cedfc8ad923cbc679049ed69fbb1baf1753617f/bundler/spec/spec_helper.rb#L1-L26

They follow that convention BUT as you can see, the tests `require bundler` as if it were external! https://guides.rubygems.org/patterns/#loading-code suggests both as OK.

I think some of the rationale could be read here: https://packaging.rubystyle.guide/#using-require-relative-from-test-to-lib https://docs.rubocop.org/rubocop-packaging/cops_packaging.html
and this seems to connect well with this comment of yours:
> Iirc, its because require_relative has limitations with where you can execute code. That mostly surfaces when you're running tests. You always have to be in the root directory, if you're in any other directory the require_relative breaks.

but from what I understand, it is exactly `require`'s problem - if you provide `require` a relative path, it is relative to the place the file was called from, while `require_relative` is always relative to the file it is in. Consider my perhaps not-so-wise example of my Chess (many files cut for brevity's sake):

```bash
├── lib
│   ├── board_navigator.rb
│   ├── board.rb
│   ├── coordinate.rb
│   ├── display.rb
│   ├── en_passant_pair.rb
│   ├── fen.rb
│   ├── fen_serializer.rb
│   ├── game_driver.rb
│   └─── game.rb
├── chess.rb
├── main.rb
└── spec
    ├── board_spec.rb
    ├── e2e
    │   └── game_driver_spec.rb
    ├── examples.txt
    ├── fen_serializer_spec.rb
    └── spec_helper.rb
```  

My `chess.rb` is a file that just `require_relatives` every single file in `lib/`. It's a shortcut I picked up from someone so I could just do `require_relative ../chess` for example in my tests, like `board_spec.rb`. With this setup, I can run tests wherever I want to. But if I change `require_relative ./lib/board.rb` in `chess.rb` to plain `require`, I can only run the test in the root of the project, because when `board_spec.rb` is loading `chess.rb`, it then gets to the `require` line and the path resolves differently, the `require` in `chess.rb` resolves from the point of view of `board_spec.rb` and fails. `require_relative`  makes sure it is resolved from the point of view of `chess.rb`.

It might be the case that `require_relative` being for in-project dependencies is going to be part of Rubocop at some point: https://github.com/rubocop/rubocop/issues/8748#issuecomment-698735044

IN SUMMARY, the tests should consider the project as an external dependency and thus use `require`, stuff in `lib` should use `require_relative`. BUT if you use `require bundler` like they do in bundler tests, you'd need to have bundler already installed... Or put differently, have `bundler.rb` somewhere in your `$LOAD_PATH` so in people's projects I guess using `require` with a relative path would be OK.

> Convention is that `require_relative` is used for your own code, while `require` is used for things outside of it, like gems that your app depend on. Now, there's couple of things that are important to know about requiring files in Ruby. <span id="namespace">First is that all required code is put into the same namespace. This means that if you have the same names for methods, modules, classes and so on they will be added together in order they were required.</span> For example:

Author: Oh and this actually means I need to change my explanations on `require` and `require_relative`, because they're wrong. Huge apologies for that.

> Convention is that `require_relative` is used for your own code, while `require` is used for things outside of it, like gems that your app depend on. Now, there's couple of things that are important to know about requiring files in Ruby. <span id="namespace">First is that all required code is put into the same namespace. This means that if you have the same names for methods, modules, classes and so on they will be added together in order they were required.</span> For example:

Reviewer 2: Thanks for sharing those links @scheals, I've got myself brushed up on this now 😄. I think I might have originally been getting things confused with the way gems load everything from lib into the LOAD_PATH. 

This all sounds solid, happy to resolve this thread.

> Convention is that `require_relative` is used for your own code, while `require` is used for things outside of it, like gems that your app depend on. Now, there's couple of things that are important to know about requiring files in Ruby. <span id="namespace">First is that all required code is put into the same namespace. This means that if you have the same names for methods, modules, classes and so on they will be added together in order they were required.</span> For example:

Reviewer: Just to chime in with another point about all of this:

I told scheals somewhere (maybe on discord?) to avoid mentioning [autoload](https://docs.ruby-lang.org/en/3.1/Kernel.html#method-i-autoload), which is another technique that's used to require code.

It's commonly used in gems to provide a kind of "lazy loading" functionality. Like taking the [noticed gem](https://github.com/excid3/noticed/blob/main/lib/noticed.rb) as an example, code is required with `autoload` such that it isn't actually required until the class or module name is used somewhere in the application code. So if an application using this gem is never delivering notifications via slack, that code is just never loaded at all.

This is good to leave out right? I figure learners at this point don't really get anything out of using it.

> Convention is that `require_relative` is used for your own code, while `require` is used for things outside of it, like gems that your app depend on. Now, there's couple of things that are important to know about requiring files in Ruby. <span id="namespace">First is that all required code is put into the same namespace. This means that if you have the same names for methods, modules, classes and so on they will be added together in order they were required.</span> For example:

Reviewer 2: Thanks @josh, no arguments from me, I think its good to leave out too. There definitely doesn't seem to be anything to from it this early, and would just muddy the water.

```rb
# all files are in the same directory for simplicity's sake

# foo.rb
def foo(bar)
  `#{bar} is awesome!`
end

# footoo.rb
def foo(bar)
  `#{bar} is awful!`
end

# my_cool_project.rb
require_relative 'foo'
require_relative 'footoo'

puts foo('Cereal')
#=> Cereal is awful!
# Since foo is defined twice, the last definition wins out.
```

To make sure code doesn't get overwritten, Rubyists wrap their code in modules which give them the benefits of a namespace:

```rb
# all files are in the same directory for simplicity's sake

# foo.rb

> module Foo

Author: I'll keep an eye on this for sure.

  def self.foo(bar)
    `#{bar} is awesome!`
  end
end
# footoo.rb
module Footoo
  def self.foo(bar)
    `#{bar} is awful!`
  end
end
# my_cool_project.rb
require_relative 'foo'
require_relative 'footoo'

puts Foo.foo('Cereal')
#=> Cereal is awesome!
puts Footoo.foo('Marmite')
#=> Marmite is awful!
puts foo('Cereal')
#=> Errors out - there's no longer a free floating foo method to use.
```

Another benefit of this approach is that you don't need to hold all the code for part of your app in one file:

> ```rb

Author: Sounds good. Since I want to keep the module-namespace example, there's no need to dive deeper into it and indeed a class example should be more understandable.

RubyGems has been part of Ruby standard library since version 1.9 and is used to get those amazing gems onto your computer. Remember that bit about `require` going through installed gems to potentially find a file you're looking for? That's the work of RubyGems. Another cool part about it? It is a gem itself! Let's give it a try.

Create a new Ruby file `main.rb` in a directory called `colorful`:

```rb
require 'colorize'

puts 'Red goes faster!'.colorize(:red)

puts "I'm blue da ba dee da ba di!".colorize(:blue)

puts "It ain't easy bein' green...".colorize(:green)
```

You're probably itching to see all those colours, so run your file with `ruby foo.rb` to see them... or rather, a LoadError. Right - you need to <span id="install-gem">install that gem first!</span> Do that with `gem install colorize` and you'll see RubyGems in action. Your system now has access to the `Colorize` gem!

> Wait, *your system* - what about others that would like to use your code? Yeah, they would need to `gem install` it as well, no big deal. But what if you have dozens of gems? How do you make sure that the versions you use are the versions others download? This sounds rather tedious. Enter: Bundler. Another gem, part of RubyGems but released independently. <span id="bundler">Bundler allows you to declare what gems your project needs - down to their version.</span> As for others, Bundler allows them to take that declaration, a simple file called `Gemfile`, and use it to install those gems in a quick `bundle install`. Since gem installs are global, <span id="bundle-exec">you need a way to run only those particular gem versions that are declared in the `Gemfile`.</span> You can do that by using `bundle exec` followed by a command that you want to execute - most likely `bundle exec ruby foo.rb`.

Reviewer 2: Few tweaks for readability - down with big paragraphs 😆 

```suggestion
Wait, *your system* - what about others who would like to use your code? Yeah, they would also need to `gem install` it - no big deal. 

But what if you have dozens of gems? How do you ensure that the versions you use are the same version others download? This sounds rather tedious. Enter: Bundler. It's another gem, part of RubyGems, but released independently. 

<span id="bundler">Bundler allows you to declare what gems your project needs - down to their version.</span> As for others, Bundler allows them to take that declaration, a simple file called `Gemfile`, and use it to install those gems in a quick `bundle install`. 

Since gem installs are global, <span id="bundle-exec">you need a way to run only those particular gem versions that are declared in the `Gemfile`.</span> You can do that by using `bundle exec` followed by a command you want to execute - most likely `bundle exec ruby foo.rb`.
```

Let's make sure whoever wants to use our script can do so:

```bash
bundle init # creates a default Gemfile in the current working directory
bundle add colorize # adds the colorize gem to the Gemfile and runs bundle install
```

This has created two files: `Gemfile` and `Gemfile.lock`. Let's take a look into both:

```rb
# Gemfile

# frozen_string_literal: true

source "https://rubygems.org"

# gem "rails"

gem "colorize", "~> 1.1"

# Gemfile.lock
GEM
  remote: https://rubygems.org/
  specs:
    colorize (1.1.0)

PLATFORMS
  ruby
  x86_64-linux # This might be different for you if you're using a different CPU and OS.

DEPENDENCIES
  colorize (~> 1.1)

BUNDLED WITH
   2.5.4
```

> <span id="gemfile">There's not much in those but as you can see,</span> the `Gemfile` has information on where to get the gems from and what gems are required. The `"~> 1.1"` is something called a version constraint, particularly one called pessimistic constraint. It relies on semantic versioning - basically, the first number is the major version, second is the minor version and a third, if it exists is the patch number. Major versions can break things from previous versions - think changing method names for example. Minor versions can add and change things but they can't break anything. Patches happen when you introduce bug fixes that don't break anything. So, if people behind a gem maintain it in line with semantic versioning, you can rely on this pessimistic constraint never letting your project have a gem version that could potentially break your app - it is equivalent to `gem "colorize", ">= 1.1", "<2.0"`.

Author: ```suggestion
<span id="gemfile">There's not much in those but as you can see,</span> the `Gemfile` has information on where to get the gems from and what gems are required. The `"~> 1.1"` is something called a version constraint, particularly one called pessimistic constraint. It relies on semantic versioning - basically, the first number is the major version, second is the minor version and a third, if it exists, is the patch number. Major versions can break things from previous versions - think changing method names for example. Minor versions can add and change things but they can't break anything. Patches happen when you introduce bug fixes that don't break anything. So, if people behind a gem maintain it in line with semantic versioning, you can rely on this pessimistic constraint never letting your project have a gem version that could potentially break your app - it is equivalent to `gem "colorize", ">= 1.1", "<2.0"`.
```

Author: Would this be a good comma to add too? 🤔

> <span id="gemfile">There's not much in those but as you can see,</span> the `Gemfile` has information on where to get the gems from and what gems are required. The `"~> 1.1"` is something called a version constraint, particularly one called pessimistic constraint. It relies on semantic versioning - basically, the first number is the major version, second is the minor version and a third, if it exists is the patch number. Major versions can break things from previous versions - think changing method names for example. Minor versions can add and change things but they can't break anything. Patches happen when you introduce bug fixes that don't break anything. So, if people behind a gem maintain it in line with semantic versioning, you can rely on this pessimistic constraint never letting your project have a gem version that could potentially break your app - it is equivalent to `gem "colorize", ">= 1.1", "<2.0"`.

Reviewer 2: To make the distinction between the different version numbers easier to visualise, we could have them broken out as bullet points:

```suggestion
<span id="gemfile">There's not much in those, but as you can see,</span> the `Gemfile` has information on where to get the gems from and what gems are required. 

The `"~> 1.1"` is a version constraint, particularly a pessimistic constraint. It relies on semantic versioning.

- The first number is the **major** version
- The second is the **minor** version
- the third, if it exists, is the **patch** number

Major versions can break things from previous versions - for example, changing method names. Minor versions can add and change things but can't break anything. Patches happen when you introduce bug fixes that don't break anything. 

So, if people behind a gem maintain it in line with semantic versioning, you can rely on this pessimistic constraint never letting your project have a gem version that could potentially break your app - it is equivalent to `gem "colorize", ">= 1.1", "<2.0"`.
```

`Gemfile.lock` has information on what was the last environment that should be able to run your app. Bundler will use it to install the same versions of gems even if `Gemfile` could potentially allow for newer versions to be installed.

### .ruby-version

> There's another one small bit that is important to give to folks that will run your code: the target Ruby version of your project. You can do it easily by running `rbenv local 3.2.2` as it creates a `.ruby-version` file with the version declared. This is recognized by many other tools to figure out what Ruby version your project is running - for example, `rbenv` will no longer use the `global` Ruby version and the `Ruby LSP` VSCode extension will also change its behavior.

Reviewer 2: ```suggestion
There's another important thing to give to folks that will run your code: the target Ruby version of your project. You can do it easily by running `rbenv local 3.2.2` as it creates a `.ruby-version` file with the version declared.

Many other tools recognize this to figure out what Ruby version your project is running - for example, `rbenv` will no longer use the `global` Ruby version and the `Ruby LSP` VSCode extension will also change its behavior.
```

### Ruby LSP in VSCode

> Earlier in the curriculum you've been instructed to choose the `Don't show again` option when `Ruby LSP` told you about not finding a lockfile - you might've also seen errors concerning `RuboCop`. In the next lesson we are going to go over `RuboCop` and how it should land in your project's `Gemfile`. After that, you are going to enjoy all the benefits of using `Ruby LSP` and its `RuboCop` integration and your projects are going to be set up like a real pro.

Reviewer 2: A few final readability tweaks:

```suggestion
Earlier in the curriculum, you were instructed to choose the `Don't show again` option when `Ruby LSP` told you about not finding a lock file - you might've also seen errors concerning `RuboCop`. In the next lesson, we will review `RuboCop` and how it should land in your project's `Gemfile`. 

After that, you will enjoy all the benefits of using `Ruby LSP` and its `RuboCop` integration, and your projects will be set up like a real pro.
```
